handlers/macrokiosk/macrokiosk.go

```diff
@@ -1,5 +1,230 @@
+	form := &moForm{}
+	err := handlers.DecodeAndValidateForm(form, r)
+	if err != nil {
+		return nil, courier.WriteAndLogRequestError(ctx, w, r, channel, err)
+	}
+
+	recipient := form.Longcode
+	sender := form.MSISDN
+	if form.Shortcode != "" {
+		recipient = form.Shortcode
+		sender = form.From
+	}
+
+	if recipient == "" || sender == "" {
+		return nil, courier.WriteAndLogRequestError(ctx, w, r, channel, fmt.Errorf("missing shortcode, longcode, from or msisdn parameters"))
+	}
+
+	if channel.Address() != recipient {
+		return nil, courier.WriteAndLogRequestError(ctx, w, r, channel, fmt.Errorf("invalid to number [%s], expecting [%s]", recipient, channel.Address()))
+	}
+
+	loc, err := time.LoadLocation("Asia/Kuala_Lumpur")
+	if err != nil {
+		return nil, courier.WriteAndLogRequestError(ctx, w, r, channel, err)
+	}
+
+	date, err := time.ParseInLocation("2006-01-0215:04:05", form.Time, loc)
+	if err != nil {
+		return nil, courier.WriteAndLogRequestError(ctx, w, r, channel, err)
+	}
+
+	// create our URN
+	urn := urns.NewTelURNForCountry(sender, channel.Country())
+
+	// build our msg
+	msg := h.Backend().NewIncomingMsg(channel, urn, form.Text).WithExternalID(form.MsgID).WithReceivedOn(date.UTC())
+
+	// and write it
+	err = h.Backend().WriteMsg(ctx, msg)
+	if err != nil {
+		return nil, err
+	}
+	courier.LogMsgReceived(r, msg)
+	return []courier.Event{msg}, h.writeReceiveSuccess(ctx, w, r, "-1") // MacroKiosk expects "-1" back for successful requests
+}
+
+func (h *handler) writeReceiveSuccess(ctx context.Context, w http.ResponseWriter, r *http.Request, responseText string) error {
+	w.WriteHeader(200)
+	_, err := fmt.Fprint(w, responseText)
+	return err
+}
+
+type mtPayload struct {
+	User   string `json:"user"`
+	Pass   string `json:"pass"`
+	To     string `json:"to"`
+	Text   string `json:"text"`
+	From   string `json:"from"`
+	ServID string `json:"servid"`
+	Type   string `json:"type"`
+}
+
+// SendMsg sends the passed in message, returning any error
+func (h *handler) SendMsg(ctx context.Context, msg courier.Msg) (courier.MsgStatus, error) {
+	username := msg.Channel().StringConfigForKey(courier.ConfigUsername, "")
+	password := msg.Channel().StringConfigForKey(courier.ConfigPassword, "")
+	servID := msg.Channel().StringConfigForKey(configMacrokioskServiceID, "")
+	senderID := msg.Channel().StringConfigForKey(configMacrokioskSenderID, "")
+	if username == "" || password == "" || servID == "" || senderID == "" {
+		return nil, fmt.Errorf("missing username, password, serviceID or senderID for MK channel")
+	}
+
+	// figure out if we need to send as unicode (encoding 5)
+	text := gsm7.ReplaceSubstitutions(handlers.GetTextAndAttachments(msg))
+	encoding := "0"
+	if !gsm7.IsValid(text) {
+		encoding = "5"
+	}
+
+	status := h.Backend().NewMsgStatusForID(msg.Channel(), msg.ID(), courier.MsgErrored)
+	parts := handlers.SplitMsg(text, maxMsgLength)
+	for _, part := range parts {
+		payload := &mtPayload{}
```

Comment:
If you are going to populate all of these at once, it might be cleaner to do:

```
payload := &mtPayload {
    From: senderID,
    ServID: servID,
   ...
}
```